src/main/java/org/folio/circulation/services/EventPublisher.java

```diff
@@ -0,0 +1,143 @@
+  public CompletableFuture<Result<CheckInProcessRecords>> publishItemCheckedInEvent(
+    CheckInProcessRecords checkInProcessRecords) {
+
+    if (checkInProcessRecords.getLoan() != null) {
+      Loan loan = checkInProcessRecords.getLoan();
+
+      JsonObject payloadJsonObject = new JsonObject();
+      write(payloadJsonObject, USER_ID_FIELD, loan.getUserId());
+      write(payloadJsonObject, LOAN_ID_FIELD, loan.getId());
+      write(payloadJsonObject, RETURN_DATE_FIELD, loan.getReturnDate());
+
+      return pubSubPublishingService.publishEvent(ITEM_CHECKED_IN.name(),
+        payloadJsonObject.encode())
+        .thenApply(r -> succeeded(checkInProcessRecords));
+    }
+    else {
+      logger.error(FAILED_TO_PUBLISH_LOG_TEMPLATE, ITEM_CHECKED_IN.name());
+    }
+
+    return completedFuture(succeeded(checkInProcessRecords));
+  }
+
+  public CompletableFuture<Result<Loan>> publishDeclaredLostEvent(Loan loan) {
+    if (loan != null) {
+      JsonObject payloadJsonObject = new JsonObject();
+      write(payloadJsonObject, USER_ID_FIELD, loan.getUserId());
+      write(payloadJsonObject, LOAN_ID_FIELD, loan.getId());
+
+      return pubSubPublishingService.publishEvent(ITEM_DECLARED_LOST.name(),
+        payloadJsonObject.encode())
+        .thenApply(r -> succeeded(loan));
+    }
+    else {
+      logger.error(FAILED_TO_PUBLISH_LOG_TEMPLATE, ITEM_DECLARED_LOST.name());
+    }
+
+    return completedFuture(succeeded(null));
+  }
+
+  private CompletableFuture<Result<Loan>> publishDueDateChangedEvent(Loan loan) {
+    if (loan != null) {
+      JsonObject payloadJsonObject = new JsonObject();
+      write(payloadJsonObject, USER_ID_FIELD, loan.getUserId());
+      write(payloadJsonObject, LOAN_ID_FIELD, loan.getId());
+      write(payloadJsonObject, DUE_DATE_FIELD, loan.getDueDate());
+      write(payloadJsonObject, DUE_DATE_CHANGED_BY_RECALL_FIELD, loan.wasDueDateChangedByRecall());
+
+      return pubSubPublishingService.publishEvent(LOAN_DUE_DATE_CHANGED.name(),
+        payloadJsonObject.encode())
+        .thenApply(r -> succeeded(loan));
+    }
+    else {
+      logger.error(FAILED_TO_PUBLISH_LOG_TEMPLATE, LOAN_DUE_DATE_CHANGED.name());
+    }
+
+    return completedFuture(succeeded(null));
+  }
+
+  public CompletableFuture<Result<LoanAndRelatedRecords>> publishDueDateChangedEvent(
+    LoanAndRelatedRecords loanAndRelatedRecords) {
+
+    if (loanAndRelatedRecords.getLoan() != null) {
+      Loan loan = loanAndRelatedRecords.getLoan();
+      publishDueDateChangedEvent(loan);
+    }
+
+    return completedFuture(succeeded(loanAndRelatedRecords));
+  }
+
+  public CompletableFuture<Result<RequestAndRelatedRecords>> publishDueDateChangedEvent(
+    RequestAndRelatedRecords requestAndRelatedRecords, Clients clients) {
+
+    LoanRepository loanRepository = new LoanRepository(clients);
+    loanRepository.findOpenLoanForRequest(requestAndRelatedRecords.getRequest())
+      .thenCompose(r -> r.after(this::publishDueDateChangedEvent));
```

Comment:
Please make sure that every endpoint that executes this code has the relevant permissions needed for `findOpenLoanForRequest`?

Reply:
Required permission is `circulation-storage.loans.collection.get`. Permission sets for all three endpoints that execute this code include it:
- `modperms.circulation.requests.item.post`
- `modperms.circulation.requests.item.put`
- `modperms.circulation.requests.item.move.post`